NIFI-3432 Handle Multiple Result Sets in ExecuteSQL

patricker opened the pull request:
This code change allows ExecuteSQL to handle queries that return multiple result sets.

One existing unit test was failing and the results didn't make sense so I adjusted it. Would be good to get a second pair of eyes on that.

I performed hand testing against MS SQL Server. In some tests I had to enable `SET NOCOUNT ON`. Results were as expected.

I put some effort into building a unit test for Derby, but it's not simple... Derby doesn't support submitting multi-statement requests in a single SQL statement, and Stored Procedures are written in Java and require registration.  I tried putting one of these together but ran into issues and decided to scrap it.

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

mattyb149 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestExecuteSQL.java:
```diff
@@ -222,7 +222,10 @@ public void testWithSqlException() throws SQLException {
+        //No incoming flow file containing a query, and an exception causes no outbound flowfile.
+        // There should be no flow files on either relatationship
+        runner.assertTransferCount(ExecuteSQL.REL_FAILURE, 0);
+        runner.assertTransferCount(ExecuteSQL.REL_FAILURE, 0);
```
These are identical, I'm guessing one should be REL_SUCCESS?

patricker replied:
Yep, it's fixed now.

mattyb149 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ExecuteSQL.java:
```diff
@@ -192,31 +192,57 @@ public void process(InputStream in) throws IOException {
+                if (resultCount > 0) {
+                    session.remove(fileToProcess);
+                } else {
+                    session.transfer(fileToProcess, REL_SUCCESS);
```
If there was an incoming flow file, and we are now writing and transferring resultSetFF, why do we transfer the original to success?  The content of incoming flow file(s) is SQL, when normal output is Avro. This might affect current flows that are only expecting Avro (or nothing). If this is a helpful addition, perhaps an "original" relationship could be added for this case.

patricker replied:
@mattyb149, Thanks for the review. In the original code `fileToProcess` would be overwritten by the results generated by the query and always outputted.  When I put in my original PR I was always dropping the original FlowFile, as it served no purpose.  Then, I happened to execute a SQL statement that returned no ResultSet... In this scenario the original FlowFile was dropped and no FlowFiles were outputted. As a result I realized I needed to keep the original FlowFile around if there were no result sets.  The above logic removes the original FlowFile under normal circumstances, `if (resultCount > 0)`.  It only keeps the original file if there are no result sets returned by the SQL.  I don't think my explanation matches up with your observations; please let me know if this helps or if there is still something wrong with the code that I just missed.

mattyb149 replied:
What was the original behavior? I would've guessed that no flow files were output if no ResultSets were returned. I'm not sure I've seen that case myself, I think usually I would get a ResultSet with no rows or something.

My concern with transferring the original to success is that it can cause flow files of different formats (SQL or Avro) to go to the same relationship. If your next processor is ConvertAvroToJSON for example, then if you execute a statement with no ResultSets, the following processor will barf on the SQL input.  If it is desired to represent the event that no ResultSets were returned, I think an "original" relationship would be a good way to go. The designer could always route both original and success to the same downstream processor if they so choose, or perhaps they want to handle "no results" some different way, so a different relationship would allow them to do that.

mattyb149 replied:
An alternative might be to create an empty Avro record (with an empty schema if that's possible) and set the attribute for number of rows to zero (and/or maybe add an attribute for "result set present" or something if needed). Whichever way you go, please add documentation where prudent (in the CapabilityDescription for example).

patricker replied:
I went out and ran the original code base. A query that returns no results causes an empty Avro schema.  I replicated this in my code, and built a test case.

mattyb149 commented:
I just noticed in the added unit test that one of the examples of a "query" that returns no ResultSets is an "UPDATE". Are you proposing that ExecuteSQL be able to handle these kinds of statements? In my experience, PutSQL is used for these kinds of statements. Do you see ExecuteSQL needing to be able to handle these as well? If so, can you describe a use case?

patricker commented:
@mattyb149 I generally use ExecuteSQL to call Stored Procedures. Unfortunately Stored Procedures are such a pain in Derby it was easier to build a test that called an update statement.

However, I have already used ExecuteSQL for Inserts in the past without issue and without modification. I didn't really even think about using PutSQL since it was a single record I was inserting for tracking metrics in an external reporting database.

patricker commented:
@mattyb149 I've rebased and squashed things down. I still don't have a good test case; as mentioned before stored procedures aren't so easy in Derby.

Would still like to see this get in.

mattyb149 commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ExecuteSQL.java:
```diff
@@ -196,44 +196,80 @@ public void process(InputStream in) throws IOException {
+
+                resultCount++;
+                // are there anymore result sets?
+                results = st.getMoreResults();
```
This should probably be in a try/catch for SQLException, some (usually younger or cross-domain) JDBC drivers may not support this method and throw an exception here. Also I think at least one driver (can't remember which one) closes the result set when the last row has been retrieved, in which case this would generate a SQLException as well. But in any case, we can just set results to false if an exception occurs and proceed from there. What do you think?

patricker commented:
@mattyb149 Sounds good. I've committed the change.

mattyb149 commented:
+1 LGTM, Ran unit tests and contrib-check, also tested with a MySQL stored procedure returning 2 ResultSets, verified two flow files were output, each with the correct row count attribute. Thanks for this improvement! Merging to master
